ddtrace/internal/_encoding.pyx

```diff
@@ -46,120 +48,230 @@ cdef inline int PyBytesLike_Check(object o):
         self.pk.buf = NULL
 
-    cdef inline object _flush_buffer(self):
-        buf = PyBytes_FromStringAndSize(self.pk.buf, self.pk.length)
-        # Reset the buffer.
-        self.pk.length = 0
-        return buf
-
-    cdef inline int _pack_number(self, object n):
-        if n is None:
-            return msgpack_pack_nil(&self.pk)
-        if PyLong_Check(n):
-            # PyInt_Check(long) is True for Python 3.
-            # So we should test long before int.
-            try:
-                if n > 0:
-                    return msgpack_pack_unsigned_long_long(&self.pk, <unsigned long long> n)
-                return msgpack_pack_long_long(&self.pk, <long long> n)
-            except OverflowError as oe:
-                if n is not self._default:
-                    return self._pack_number(self._default)
-                raise OverflowError("Integer value out of range")
-
-        elif PyInt_Check(n):
-            return msgpack_pack_long(&self.pk, <long> n)
-
-        elif PyFloat_Check(n):
-            return msgpack_pack_double(&self.pk, <double> n)
-
-        raise TypeError("Unhandled numeric type: %r" % type(n))
-
-    cdef inline int _pack_text(self, object text):
+    def __len__(self):  # TODO: Use a better name?
+        return self._count
+
+    cpdef _decode(self, data):
+        import msgpack
+        if msgpack.version[:2] < (0, 6):
+            return msgpack.unpackb(data)
+        return msgpack.unpackb(data, raw=True)
+
+    cdef _reset_buffer(self):
+        self._count = 0
+        self.pk.length = MSGPACK_ARRAY_LENGTH_PREFIX_SIZE  # Leave room for array length prefix
+
+    cpdef encode(self):
+        if not self._count:
+            return None
+
+        return self.flush()
+
+    cdef inline int _update_array_len(self):
+        """Update traces array size prefix"""
+        cdef int offset = MSGPACK_ARRAY_LENGTH_PREFIX_SIZE - array_prefix_size(self._count)
+        cdef int old_pos = self.pk.length
+
+        with self._lock:
+            self.pk.length = offset
+            msgpack_pack_array(&self.pk, self._count)
+            self.pk.length = old_pos
+            return offset
+
+    cpdef get_bytes(self):
+        """Return internal buffer contents as bytes object"""
+        cdef int offset = self._update_array_len()
+        with self._lock:
+            return PyBytes_FromStringAndSize(self.pk.buf + offset, self.pk.length - offset)
+
+    cpdef char * get_buffer(self):
+        """Return internal buffer."""
+        return self.pk.buf + self._update_array_len()
+
+    cdef inline int _pack_trace(self, list trace):
+        cdef int ret
         cdef Py_ssize_t L
+
+        L = len(trace)
+        if L > ITEM_LIMIT:
+            raise ValueError("list is too large")
+
+        ret = msgpack_pack_array(&self.pk, L)
+        if ret != 0: raise RuntimeError("Couldn't pack trace")
+
+        if L > 0 and trace[0].context is not None:
+            dd_origin = trace[0].context.dd_origin
```

Comment:
Same here. We should see if it's possible to unbox this once and use `msgpack_pack_raw*` instead of `pack_text`. My only concern is with `pack_span`, but given that here it is still `cdef`ined we should be good.

Reply:
you mean something like:

```suggestion
            dd_origin = <char*> trace[0].context.dd_origin
```

?

pass around the c string rather than pulling it from the PyUnicode every time?

Reply:
yes!

Reply:
I'll pull out the dd_origin improvement changes into a separate PR so we can keep this one as small as possible :)